userspace/engine/rule_loader.cpp

```diff
@@ -0,0 +1,1098 @@
+/*
+Copyright (C) 2022 The Falco Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "falco_engine.h"
+#include "rule_loader.h"
+#include "filter_macro_resolver.h"
+
+#define MAX_VISIBILITY	  ((uint32_t) -1)
+#define THROW(cond, err)	{ if (cond) { throw falco_exception(err); } }
+
+static string s_container_info_fmt = "%container.info";
+static string s_default_extra_fmt  = "%container.name (id=%container.id)";
+
+using namespace std;
+using namespace libsinsp::filter;
+
+string ctxerr(std::string ctx, std::string e)
+{
+	e += "\n---\n";
+	e += trim(ctx);
+	e += "\n---";
+	return e;
+}
+
+// todo(jasondellaluce): this breaks string escaping in lists and exceptions
+static void quote_item(string& e)
+{
+	if (e.find(" ") != string::npos && e[0] != '"' && e[0] != '\'')
+	{
+		e = '"' + e + '"';
+	}
+}
+
+static void paren_item(std::string& e)
+{
+	if(e[0] != '(')
+	{
+		e = '(' + e + ')';
+	}
+}
+
+static bool is_field_defined(falco_engine *engine, string source, string field)
+{
+	auto factory = engine->get_filter_factory(source);
+	if(factory)
+	{
+		auto *chk = factory->new_filtercheck(field.c_str());
+		if (chk)
+		{
+			delete(chk);
+			return true;
+		}
+	}
+	return false;
+}
+
+// todo: this should be in libsinsp
+static bool is_operator_defined(std::string op)
+{
+	static vector<string> ops = {"=", "==", "!=", "<=", ">=", "<", ">",
+		"contains", "icontains", "bcontains", "glob", "bstartswith",
+		"startswith", "endswith", "in", "intersects", "pmatch"};
+	return find(ops.begin(), ops.end(), op) != ops.end();
+}
+
+// todo: this should be in libsinsp
+static bool is_operator_for_list(std::string op)
+{
+	return op == "in" || op == "intersects" || op == "pmatch";
+}
+
+static bool is_format_valid(falco_engine* e, string src, string fmt, string& err)
+{
+	try
+	{
+		shared_ptr<gen_event_formatter> formatter;
+		formatter = e->create_formatter(src, fmt);
+		return true;
+	}
+	catch(exception &e)
+	{
+		err = e.what();
+		return false;
+	}
+}
+
+static string yaml_format_object(
```

Comment:
Yeah this is the point. I just checked with YAML::Emitter (which is also what YAML::Dump uses internally), and all the comments/indentation is lost. Since the goal of the yaml context is to give users a pointer to a YAML file portion as-is, I think it's better to keep this approach and maybe think about something better in the future.

Reply:
Yeah that's too bad. The go yaml parser (https://github.com/go-yaml/yaml), v3, actually *does* preserve all that stuff, which is a fantastic update. It didn't seem like yaml-cpp or biojppm/rapidyaml had the same feature, though.